deepcave/utils/symbolic_regression.py

```diff
@@ -0,0 +1,110 @@
+
+
+def convert_symb(symb, n_decimals: int = None, hp_names: list = None) -> sympy.core.expr:
+    """
+    Convert a fitted symbolic regression to a simplified and potentially rounded mathematical expression.
+    Warning: eval is used in this function, thus it should not be used on unsanitized input (see
+    https://docs.sympy.org/latest/modules/core.html?highlight=eval#module-sympy.core.sympify).
+
+    Parameters
+    ----------
+    symb: Fitted symbolic regressor to find a simplified expression for.
+    n_decimals: If set, round floats in the expression to this number of decimals.
+    hp_names: If set, replace X0 and X1 in the expression by the names given.
+
+    Returns
+    -------
+    symb_conv: Converted mathematical expression.
+    """
+
+    # sqrt is protected function in gplearn, always returning sqrt(abs(x))
+    sqrt_pos = []
+    prev_sqrt_inserts = 0
+    for i, f in enumerate(symb._program.program):
+        if isinstance(f, functions._Function) and f.name == "sqrt":
+            sqrt_pos.append(i)
+    for i in sqrt_pos:
+        symb._program.program.insert(i + prev_sqrt_inserts + 1, functions.abs1)
+        prev_sqrt_inserts += 1
+
+    # log is protected function in gplearn, always returning sqrt(abs(x))
+    log_pos = []
+    prev_log_inserts = 0
+    for i, f in enumerate(symb._program.program):
+        if isinstance(f, functions._Function) and f.name == "log":
+            log_pos.append(i)
+    for i in log_pos:
+        symb._program.program.insert(i + prev_log_inserts + 1, functions.abs1)
+        prev_log_inserts += 1
+
+    symb_str = str(symb._program)
+
+    converter = {
+        "sub": lambda x, y: x - y,
+        "div": lambda x, y: x / y,
+        "mul": lambda x, y: x * y,
+        "add": lambda x, y: x + y,
+        "neg": lambda x: -x,
+        "pow": lambda x, y: x**y,
+    }
+
+    # Abort conversion for very long programs, as they take too much time or do not finish at all.
+    if symb._program.length_ > 300:
+        return symb_str
+
+    # Convert formula string to SymPy object
+    symb_conv = sympy.sympify(symb_str.replace("[", "").replace("]", ""), locals=converter)
+
+    # Replace variable names in formula by hyperparameter names
+    if hp_names is not None:
+        if len(hp_names) == 1:
+            X0, hp0 = sympy.symbols(f"X0 {hp_names[0]}")
+            symb_conv = symb_conv.subs(X0, hp0)
+        elif len(hp_names) == 2:
+            X0, hp0, X1, hp1 = sympy.symbols(f"X0 {hp_names[0]} X1 {hp_names[1]}")
+            symb_conv = symb_conv.subs(X0, hp0)
+            symb_conv = symb_conv.subs(X1, hp1)
+        else:
+            raise ValueError(
+                f"Numer of hyperparameters to be explained by symbolic explanations must not "
+                f"be larger than 2"
+            )
+
+    # Simplify the formula
+    try:
+        # Simplification can fail in some cases. If so, use the unsimplified version.
+        symb_simpl = sympy.simplify(symb_conv)
+    except:
```

Comment:
Is there a specific exception that could be caught here instead?

Reply:
I unfortunately didn't track that and cannot find an example now, thus I just changed it to actually print the exception (so if I should see one at some point, I can edit it).